services/user/Tokens/src/RTokens.cpp

```diff
@@ -118,81 +130,52 @@ struct TokenQuery
 
    /// Provides a way to query all balances for a given user.
    /// Further filtering can be done to restrict the results by balance.
-   auto userBalances(AccountNumber         user,
-                     optional<uint32_t>    first,
-                     optional<uint32_t>    last,
-                     optional<std::string> before,
-                     optional<std::string> after) const
+   auto userBalances(AccountNumber user) const
    {
-      vector<TokenBalance> balances;
-
       auto tokenTypeIdx = tokenService.open<TokenTable>().getIndex<0>();
       auto balanceIdx   = tokenService.open<BalanceTable>().getIndex<0>().subindex(
           SplitKey<TID, AccountNumber>{user});
 
-      // Balances of this user
-      for (auto balance : balanceIdx)
-      {
-         auto tokenOpt = tokenTypeIdx.get(balance.key.tokenId);
-         check(tokenOpt.has_value(), "Invalid token type");
-         balances.push_back(TokenBalance{tokenOpt->symbolId, balance.key.tokenId,
-                                         tokenOpt->precision, balance.balance});
-      }
-
-      return makeVirtualConnection<
-          Connection<TokenBalance, "UserBalanceConnection", "UserBalanceEdge">, TokenBalance, SID>(
-          balances, {std::nullopt}, {std::nullopt}, {std::nullopt}, {std::nullopt}, first, last,
-          before, after);
+      return TransformedConnection{balanceIdx,
+                                   [tokenTypeIdx = std::move(tokenTypeIdx)](auto&& balance)
+                                   {
+                                      auto tokenOpt = tokenTypeIdx.get(balance.key.tokenId);
+                                      check(tokenOpt.has_value(), "Invalid token type");
+                                      return TokenBalance{tokenOpt->symbolId, balance.key.tokenId,
+                                                          tokenOpt->precision, balance.balance};
+                                   }};
    }
 
-   auto userCredits(AccountNumber         user,
-                    optional<uint32_t>    first,
-                    optional<uint32_t>    last,
-                    optional<std::string> before,
-                    optional<std::string> after) const
+   auto userCredits(AccountNumber user) const
    {
-      vector<Credit> credits;
-
       auto tokenTypeIdx = tokenService.open<TokenTable>().getIndex<0>();
       auto creditIdx    = tokenService.open<SharedBalanceTable>().getIndex<0>().subindex(
-          SplitKey<std::tuple<TID, AccountNumber>, AccountNumber>{user});
```

Comment:
Why did you change this to BalanceKey? It is the same types, but BalanceKey is related to the BalanceTable, not the SharedBalanceTable

Reply:
The key needs to be a struct with named members, not a tuple, so it can be passed as an argument (ge/gt/lt/le).